c++/src/capnp/serialize-async.c++

```diff
@@ -502,4 +506,334 @@ kj::Promise<MessageReaderAndFds> MessageStream::readMessage(
+    BufferedStreamOptions options) {
+  return kj::heap<BufferedAsyncMessageStream>(ioStream, options);
+}
+
+kj::Own<MessageStream> makeBufferedStream(kj::AsyncCapabilityStream& capStream,
+    BufferedStreamOptions options) {
+  return kj::heap<BufferedAsyncMessageStream>(capStream, options);
+}
+
+kj::Promise<BufferedAsyncMessageStream::ReadResult>
+    BufferedAsyncMessageStream::tryRead(kj::AsyncIoStream& stream,
+      void* start, size_t minBytes, size_t maxBytes) {
+  return stream.tryRead(start, minBytes, maxBytes).then(
+      [](size_t n) mutable {
+    return ReadResult{ n, 0, nullptr};
+  });
+}
+
+kj::Promise<BufferedAsyncMessageStream::ReadResult>
+    BufferedAsyncMessageStream::tryRead(kj::AsyncCapabilityStream& stream,
+      void* start, size_t minBytes, size_t maxBytes) {
+  auto fdSpace = kj::heapArray<kj::AutoCloseFd>(options.maxFdsPerMessage);
+  return stream.tryReadWithFds(start, minBytes, maxBytes, fdSpace.begin(),
+      options.maxFdsPerMessage).then(
+        [fdSpace = kj::mv(fdSpace)](kj::AsyncCapabilityStream::ReadResult result) mutable {
+    return ReadResult{ result.byteCount, result.capCount, kj::mv(fdSpace) };
+  });
+}
+
+kj::Promise<kj::Maybe<MessageReaderAndFds>> BufferedAsyncMessageStream::tryReadMessage(
+    kj::ArrayPtr<kj::AutoCloseFd> passedInFdSpace,
+    ReaderOptions readerOptions, kj::ArrayPtr<word> scratchSpace) {
+  KJ_IF_MAYBE(m, tryGetBufferedMessage(passedInFdSpace)) {
+    return kj::Maybe<MessageReaderAndFds>(kj::mv(*m));
+  }
+
+  auto prefixSize = 0;
+  kj::ArrayPtr<word> buffer;
+  kj::Array<word> ownBuffer;
+  void* start;
+  size_t minBytes;
+  size_t maxBytes;
+
+  KJ_IF_MAYBE(i, incompleteRead) {
+    // We take the incomplete read and put it as a prefix in the readBuffer.
+
+    KJ_ASSERT(i->bytes.size() < i->expectedSizeInWords * sizeof(word));
+    prefixSize = i->bytes.size();
+
+    // The buffer needs to be at least as big as the expected message size, so if the message is
+    // larger than defaultReadSizeInWords, we need to use that
+    auto bufferSizeInWords = kj::max(i->expectedSizeInWords, options.defaultReadSizeInWords);
+    if (scratchSpace.size() >= bufferSizeInWords) {
+      buffer = scratchSpace;
+    } else {
+      ownBuffer = kj::heapArray<word>(bufferSizeInWords);
+      buffer = ownBuffer.asPtr();
+    }
+
+    memcpy(static_cast<void*>(buffer.begin()), static_cast<void*>(i->bytes.begin()), prefixSize);
+
+    // We want to start the read after the prefix, and so we must adjust start, minBytes, and maxBytes
+    // to account for the prefix.
+    start = buffer.asBytes().begin() + prefixSize;
+    minBytes = i->expectedSizeInWords * sizeof(word) - prefixSize;
+    maxBytes = bufferSizeInWords * sizeof(word) - prefixSize;
+  } else {
+    if (scratchSpace.size() >= options.defaultReadSizeInWords) {
+      buffer = scratchSpace;
+    } else {
+      ownBuffer = kj::heapArray<word>(options.defaultReadSizeInWords);
```

Comment:
> better for processor cache hit rate

Copying all the messages out of the read buffer into separate message objects proactively seems worse for the cache to me. In this case, those separate objects will be cold in cache by the time you get to actually processing them. And since those objects are not contiguous chunks of memory, but rather a few different C++ objects malloc()ed in different places, there would be more cache misses in processing them.

That said, performance at this level is a mysterious thing and hard to predict confidently. We'd need to test to be sure.

> What if we returned MessageReaders that pointed into the read buffer, and said that any call to tryReadMessage would then invalidate any previously returned MessageReaders

Hmm. Well, this would make things slower if you _weren't_ using buffered reads, as the copies are unnecessary in that case. It also potentially makes things slower in the case of a `Call` or `Return` message that is much larger than the buffer; in this case, presumably the `MessageReader` would already read directly into a buffer dedicated to the message, and so making a copy later is not necessary.

One hacky optimization we could do here -- without an interface change -- is we could say that if a single message in the buffer occupies more than n% of the buffer, we're willing to transfer ownership of the buffer to it (or refcounted, if there is more than one), rather than make a copy. This wastes a bit of memory but saves copies of larger messages. Small messages, meanwhile, are probably not a big deal to copy.

With that said, I could be convinced to accept an interface change if it shows big wins. But I think I'd prefer a change where the RPC system has a way to request ownership of the message, and the MessageReader then decides whether or not to make a copy, as appropriate for its implementation.